Signal/src/call/CallService.swift

```diff
@@ -98,6 +98,108 @@ protocol CallServiceObserver: class {
                               remoteVideoTrack: RTCVideoTrack?)
 }
 
+// Gather all per-call state in one place.
+private class SignalCallData: NSObject {
+    public let call: SignalCall
+
+    // Used to coordinate promises across delegate methods
+    let fulfillCallConnectedPromise: (() -> Void)
+    let rejectCallConnectedPromise: ((Error) -> Void)
+    let callConnectedPromise: Promise<Void>
+
+    // Used to ensure any received ICE messages wait until the peer connection client is set up.
+    let fulfillPeerConnectionClientPromise: (() -> Void)
+    let rejectPeerConnectionClientPromise: ((Error) -> Void)
+    let peerConnectionClientPromise: Promise<Void>
+
+    // Used to ensure CallOffer was sent before sending any ICE updates.
+    let fulfillReadyToSendIceUpdatesPromise: (() -> Void)
+    let rejectReadyToSendIceUpdatesPromise: ((Error) -> Void)
+    let readyToSendIceUpdatesPromise: Promise<Void>
+
+    weak var localVideoTrack: RTCVideoTrack? {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            Logger.info("\(self.logTag) \(#function)")
+        }
+    }
+
+    weak var remoteVideoTrack: RTCVideoTrack? {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            Logger.info("\(self.logTag) \(#function)")
+        }
+    }
+
+    var isRemoteVideoEnabled = false {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            Logger.info("\(self.logTag) \(#function): \(isRemoteVideoEnabled)")
+        }
+    }
+
+    var peerConnectionClient: PeerConnectionClient? {
+        didSet {
+            SwiftAssertIsOnMainThread(#function)
+
+            Logger.debug("\(self.logTag) .peerConnectionClient setter: \(oldValue != nil) -> \(peerConnectionClient != nil) \(String(describing: peerConnectionClient))")
+        }
+    }
+
+    required init(call: SignalCall) {
+        self.call = call
+
+        let (callConnectedPromise, fulfillCallConnectedPromise, rejectCallConnectedPromise) = Promise<Void>.pending()
```

Comment:
I really like having all these per call callbacks centralized like this. Good work!